pkg/smart_wallet/init_code.go

```diff
@@ -0,0 +1,157 @@
+	initCode, err := GetInitCode(provider, config, owner, index)
+	if err != nil {
+		return common.Address{}, fmt.Errorf("failed to build initCode middleware: %w", err)
+	}
+
+	// calculate the smart wallet address that will be generated by the entry point
+	// See https://github.com/eth-infinitism/account-abstraction/blob/v0.6.0/contracts/core/EntryPoint.sol#L356
+	getSenderAddressData, err := entryPointABI.Pack("getSenderAddress", initCode)
+	if err != nil {
+		return common.Address{}, fmt.Errorf("failed to pack getSenderAddress data: %w", err)
+	}
+
+	msg := ethereum.CallMsg{
+		To:   &entryPointAddress,
+		Data: getSenderAddressData,
+	}
+
+	_, err = provider.CallContract(ctx, msg, nil)
+	if err == nil {
+		panic(fmt.Errorf("'getSenderAddress' call returned no error, but expected one"))
+	}
+
+	var scError rpc.DataError
+	if ok := errors.As(err, &scError); !ok {
+		panic(fmt.Errorf("unexpected error type '%T' containing message %w)", err, err))
+	}
+	errorData := scError.ErrorData().(string)
+
+	senderAddressResultError, ok := entryPointABI.Errors["SenderAddressResult"]
+	if !ok {
+		panic(fmt.Errorf("ABI does not contain 'SenderAddressResult' error"))
+	}
+
+	// check if the error signature is correct
+	if id := senderAddressResultError.ID.String(); errorData[0:10] != id[0:10] {
+		panic(fmt.Errorf("'getSenderAddress' unexpected error signature: %s", errorData[0:10]))
+	}
+
+	// check if the error data has the correct length
+	if len(errorData) < 74 {
+		panic(fmt.Errorf("'getSenderAddress' revert data expected to have lenght of 74, but got: %d", len(errorData)))
+	}
+
+	swAddress := common.HexToAddress(errorData[34:])
+	if swAddress == (common.Address{}) {
+		panic(fmt.Errorf("'getSenderAddress' returned zero address"))
+	}
+
+	return swAddress, nil
+}
+
+func GetInitCode(provider *ethclient.Client, smartWalletConfig Config, ownerAddress common.Address, index decimal.Decimal) ([]byte, error) {
+	var initCode []byte
+	var err error
+
+	switch typ := *smartWalletConfig.Type; typ {
+	case SimpleAccountType:
+		return nil, fmt.Errorf("%w: %s", ErrSmartWalletNotSupported, typ)
+	case BiconomyType: // not tested
+		initCode, err = GetBiconomyInitCode(ownerAddress, index, smartWalletConfig.Factory, smartWalletConfig.ECDSAValidator)
+	case KernelType:
+		initCode, err = GetKernelInitCode(ownerAddress, index, smartWalletConfig.Factory, smartWalletConfig.Logic, smartWalletConfig.ECDSAValidator)
+	default:
+		return nil, fmt.Errorf("unknown smart wallet type: %s", typ)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get init code: %w", err)
+	}
+
+	return initCode, nil
+}
+
+// getKernelInitCode returns a middleware that sets the init code
+// for a Zerodev Kernel smart account. The init code deploys
+// a smart account if it is not already deployed.
+func GetKernelInitCode(owner common.Address, index decimal.Decimal, factory, accountLogic, ecdsaValidator common.Address) ([]byte, error) {
```

Comment:
replace all unnecessary panics with error return
P.S. you do not validate inputs, so not internal error is still possible
And make index non-decimal, it makes no sense

Reply:
2b120d650ab1e4b8f991754ca18ef50b8e7177f5

Reply:
Also, I will change `index decimal.Decimal` to `index *big.Int` in the following PRs